File: typescript/tests/legal_board.test.ts
// legal_board.test.ts
// Verifies the CircleAI.Legal port: matters (open/close/active ordering),
// contracts expiring before a date, upcoming deadlines, and clause-by-tag.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryLegalBoard,
  LegalDomainContext,
  matter,
  contract,
  legalDeadline,
  clause,
} from "../src/legal/index";

describe("InMemoryLegalBoard — matters", () => {
  it("opens, closes, and lists active matters newest-first", () => {
    const b = new InMemoryLegalBoard();
    b.open(matter("m1", "Older", "ZA", "C1", new Date("2026-01-01T00:00:00Z"), true));
    b.open(matter("m2", "Newer", "ZA", "C2", new Date("2026-06-01T00:00:00Z"), true));
    b.open(matter("m3", "Closed", "ZA", "C3", new Date("2026-05-01T00:00:00Z"), true));
    b.close("m3");
    assert.equal(b.getMatter("m3")?.open, false);
    assert.deepEqual(
      b.activeMatters.map((m) => m.matterId),
      ["m2", "m1"],
    );
  });

  it("closing an unknown matter throws", () => {
    const b = new InMemoryLegalBoard();
    assert.throws(() => b.close("ghost"), /Unknown matter ghost/);
  });
});

describe("InMemoryLegalBoard — contracts", () => {
  it("contractsExpiringBefore includes only dated contracts <= date, sorted asc", () => {
    const b = new InMemoryLegalBoard();
    b.addContract(contract("c1", "m1", "A", new Date("2026-01-01"), new Date("2026-12-31"), ["X"]));
    b.addContract(contract("c2", "m1", "B", new Date("2026-01-01"), new Date("2026-03-31"), ["Y"]));
    b.addContract(contract("c3", "m1", "Perp", new Date("2026-01-01"), null, ["Z"]));
    b.addContract(contract("c4", "m1", "Late", new Date("2026-01-01"), new Date("2027-06-30"), ["W"]));
    const hits = b.contractsExpiringBefore(new Date("2026-12-31T23:59:59Z"));
    assert.deepEqual(
      hits.map((c) => c.contractId),
      ["c2", "c1"],
    );
  });
});

describe("InMemoryLegalBoard — deadlines", () => {
  it("upcomingDeadlines returns DueOn >= now, sorted ascending", () => {
    const b = new InMemoryLegalBoard();
    b.add(legalDeadline("d1", "m1", "past", new Date("2026-01-01")));
    b.add(legalDeadline("d2", "m1", "soon", new Date("2026-08-01")));
    b.add(legalDeadline("d3", "m1", "later", new Date("2026-12-01")));
    const up = b.upcomingDeadlines(new Date("2026-07-01"));
    assert.deepEqual(
      up.map((d) => d.deadlineId),
      ["d2", "d3"],
    );
  });
});

describe("InMemoryLegalBoard — clauses", () => {
  it("clausesByTag matches case-insensitively", () => {
    const b = new InMemoryLegalBoard();
    b.addClause(clause("cl1", "Indemnity", "…", ["Liability", "Risk"]));
    b.addClause(clause("cl2", "Warranty", "…", ["risk"]));
    b.addClause(clause("cl3", "Term", "…", ["Duration"]));
    assert.deepEqual(
      b.clausesByTag("RISK").map((c) => c.clauseId).sort(),
      ["cl1", "cl2"],
    );
  });

  it("a blank tag throws", () => {
    const b = new InMemoryLegalBoard();
    assert.throws(() => b.clausesByTag("   "), /tag required/);
  });
});

describe("LegalDomainContext", () => {
  it("exposes prompt + compliance + tools", () => {
    assert.ok(LegalDomainContext.systemPromptSnippet.includes("[DOMAIN: Legal]"));
    assert.deepEqual(LegalDomainContext.complianceFlags, [
      "Legal_Practice_Act_28_2014",
      "Attorneys_Act",
      "POPIA",
      "Professional_Legal_Privilege",
    ]);
    assert.deepEqual(LegalDomainContext.suggestedTools, ["legal_research", "document_editor", "contract_analyser"]);
  });
});

File: go/tests/legal_board_test.go
// legal_board_test.go
//
// Verifies the CircleAI.Legal port (legal_board.go): matter open/close/active
// ordering, contract expiry filtering + ordering, deadline upcoming filtering,
// and clause tag lookup (case-insensitive) with blank-tag error.

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func legalTime(y int, m time.Month, d int) time.Time {
	return time.Date(y, m, d, 0, 0, 0, 0, time.UTC)
}

func TestLegal_MattersOpenCloseActive(t *testing.T) {
	b := circleai.NewInMemoryLegalBoard()
	b.Open(circleai.Matter{MatterId: "m1", Title: "One", OpenedAtUtc: legalTime(2026, 1, 1), Open: true})
	b.Open(circleai.Matter{MatterId: "m2", Title: "Two", OpenedAtUtc: legalTime(2026, 3, 1), Open: true})
	b.Open(circleai.Matter{MatterId: "m3", Title: "Three", OpenedAtUtc: legalTime(2026, 2, 1), Open: true})

	active := b.ActiveMatters()
	if len(active) != 3 || active[0].MatterId != "m2" || active[1].MatterId != "m3" || active[2].MatterId != "m1" {
		t.Fatalf("active matters desc-by-open failed: %+v", active)
	}
	if err := b.Close("m2"); err != nil {
		t.Fatalf("close: %v", err)
	}
	if active = b.ActiveMatters(); len(active) != 2 {
		t.Fatalf("after close want 2 active, got %d", len(active))
	}
	if m, ok := b.GetMatter("m2"); !ok || m.Open {
		t.Fatalf("m2 should be closed: %+v ok=%v", m, ok)
	}
	if err := b.Close("ghost"); err == nil {
		t.Fatalf("closing unknown matter must error")
	}
}

func TestLegal_ContractsExpiringBefore(t *testing.T) {
	b := circleai.NewInMemoryLegalBoard()
	exp1 := legalTime(2026, 6, 1)
	exp2 := legalTime(2026, 8, 1)
	b.AddContract(circleai.Contract{ContractId: "c1", MatterId: "m1", Title: "Early", EffectiveDate: legalTime(2026, 1, 1), ExpiryDate: &exp1, Counterparties: []string{"X"}})
	b.AddContract(circleai.Contract{ContractId: "c2", MatterId: "m1", Title: "Late", EffectiveDate: legalTime(2026, 1, 1), ExpiryDate: &exp2})
	b.AddContract(circleai.Contract{ContractId: "c3", MatterId: "m1", Title: "Perpetual", EffectiveDate: legalTime(2026, 1, 1), ExpiryDate: nil})

	before := b.ContractsExpiringBefore(legalTime(2026, 7, 1))
	if len(before) != 1 || before[0].ContractId != "c1" {
		t.Fatalf("expiring-before failed: %+v", before)
	}
	// On-or-before is inclusive of the exact date; both c1 and c2 by Sep.
	all := b.ContractsExpiringBefore(legalTime(2026, 9, 1))
	if len(all) != 2 || all[0].ContractId != "c1" || all[1].ContractId != "c2" {
		t.Fatalf("inclusive expiring order failed: %+v", all)
	}
}

func TestLegal_ContractCounterpartiesCopied(t *testing.T) {
	b := circleai.NewInMemoryLegalBoard()
	exp := legalTime(2026, 6, 1)
	parties := []string{"Acme"}
	b.AddContract(circleai.Contract{ContractId: "c1", MatterId: "m1", ExpiryDate: &exp, Counterparties: parties})
	parties[0] = "MUTATED"
	got := b.ContractsExpiringBefore(legalTime(2027, 1, 1))
	if got[0].Counterparties[0] != "Acme" {
		t.Fatalf("counterparties not defensively copied: %v", got[0].Counterparties)
	}
}

func TestLegal_UpcomingDeadlines(t *testing.T) {
	b := circleai.NewInMemoryLegalBoard()
	b.Add(circleai.LegalDeadline{DeadlineId: "d1", MatterId: "m1", Description: "past", DueOn: legalTime(2026, 1, 1)})
	b.Add(circleai.LegalDeadline{DeadlineId: "d2", MatterId: "m1", Description: "soon", DueOn: legalTime(2026, 7, 15)})
	b.Add(circleai.LegalDeadline{DeadlineId: "d3", MatterId: "m1", Description: "later", DueOn: legalTime(2026, 9, 1)})
	up := b.UpcomingDeadlines(legalTime(2026, 7, 1))
	if len(up) != 2 || up[0].DeadlineId != "d2" || up[1].DeadlineId != "d3" {
		t.Fatalf("upcoming deadlines failed: %+v", up)
	}
}

func TestLegal_ClausesByTag(t *testing.T) {
	b := circleai.NewInMemoryLegalBoard()
	b.AddClause(circleai.Clause{ClauseId: "cl1", Title: "Indemnity", Body: "...", Tags: []string{"Risk", "Liability"}})
	b.AddClause(circleai.Clause{ClauseId: "cl2", Title: "Term", Body: "...", Tags: []string{"General"}})
	b.AddClause(circleai.Clause{ClauseId: "cl3", Title: "Cap", Body: "...", Tags: []string{"liability"}})

	hits, err := b.ClausesByTag("LIABILITY")
	if err != nil {
		t.Fatalf("by tag: %v", err)
	}
	if len(hits) != 2 || hits[0].ClauseId != "cl1" || hits[1].ClauseId != "cl3" {
		t.Fatalf("case-insensitive tag match failed: %+v", hits)
	}
	if _, err := b.ClausesByTag("  "); err == nil {
		t.Fatalf("blank tag must error")
	}
}
